synapseclient/core/utils.py

```diff
@@ -342,12 +344,39 @@ def is_same_base_url(url1: str, url2: str) -> bool:
 def is_synapse_id_str(obj):
     """If the input is a Synapse ID return it, otherwise return None"""
     if isinstance(obj, str):
-        m = re.match(r"(syn\d+$)", obj)
+        m = re.match(r"(syn\d+(\.\d+)?$)", obj)
         if m:
             return m.group(1)
     return None
 
 
+def get_synid_and_version(obj):
+    """Extract the Synapse ID and version number from input entity"""
+
+    if isinstance(obj, str):
+        synapse_id_and_version = is_synapse_id_str(obj)
+        m = re.match(r"(syn\d+)(?:\.(\d+))?", synapse_id_and_version)
+        id = m.group(1)
+        version = m.group(2)
+    elif isinstance(obj, synapseclient.Entity):
+        id = id_of(obj)
+        version = None
+        if "versionNumber" in obj:
+            version = obj["versionNumber"]
+
+    return id, version
+
+
+def get_uri(method, entity_id, version=None):
```

Comment:
I don't think what you are trying to do is going to work here. Not every endpoint in the Synapse REST API can be accessed with a version.

Take the benefactor for example.

Works:
```
curl --location 'https://repo-prod.prod.sagebase.org/repo/v1/entity/syn53362132/benefactor' \
--header 'Authorization: Bearer '
```

Does not work because the endpoint does not exist:
```
curl --location 'https://repo-prod.prod.sagebase.org/repo/v1/entity/syn53362132/version/1/benefactor' \
--header 'Authorization: Bearer '
```

Reply:
I see, yep I wrote this when I was planning for the larger `.version` support across the whole codebase. I'll retract this for now, but I think later maybe it's worth considering having a `get_uri` function in utils that will generate the URI for us on these calls, and have the endpoint be a kwarg the way it is now, and trigger building the URI with `version` if two conditions are met: the input entity (str or object) has a versionNumber, and the endpoint is one that accepts versionNumber, which can be pulled from a list.

This way any bugs caused by broken URIs can all be traced back to that one function.

Let me know your thoughts

Reply:
Sounds good. I might consider naming the function when you do implement it, something like `get_versionable_entity_uri`. `get_uri` is a bit too generic for the specific logic the function will hold.

Reply:
Makes sense. I'll add this convo to the general `.version` notation ticket [here](https://sagebionetworks.jira.com/browse/SYNPY-699).